File: publish/project/TsScripts/System/unit/SpecialAnimationPlayer.ts
export class SpecialAnimationPlayer {
    static createSpecialAnimation(obj: UnityEngine.GameObject): boolean {
        let animationPlayer = obj.GetComponent(Game.SpecialAnimationPlayer.GetType()) as Game.SpecialAnimationPlayer;
        if (animationPlayer == null) {
            return false;
        }
        let func = SpecialAnimationPlayer[animationPlayer.functionName];
        if (func == null) {
            uts.logError("函数不存在：" + animationPlayer.functionName);
            return false;
        }
        func(animationPlayer);
        return true;
    }
    static ActiveChildren(animationPlayer: Game.SpecialAnimationPlayer) {
        let transform = animationPlayer.transform;
        var count = transform.childCount;
        if (count == 0) {
            return;
        }
        let f = animationPlayer.GetFloat(0) * 1000;
        for (let i = 0; i < count; i++) {
            let gameObject = transform.GetChild(i).gameObject;
            gameObject.SetActive(false);
            new Game.Timer("skill", f * i, 1, delegate(this, SpecialAnimationPlayer.onLateActiveChildren, gameObject));
        }
    }
    private static onLateActiveChildren(timer: Game.Timer, obj: UnityEngine.GameObject) {
        obj.SetActive(true);
    }

    static Play() {

    }
    static RandomActiveChildren(animationPlayer: Game.SpecialAnimationPlayer) {
        let transformList = [];
        let transform = animationPlayer.transform;
        var count = transform.childCount;
        if (count == 0) {
            return;
        }
        for (let i = 0; i < count; i++) {
            let gameObject = transform.GetChild(i).gameObject;
            gameObject.SetActive(false);
            transformList.push(gameObject);
        }
        let f = animationPlayer.GetFloat(0) * 1000;
        let i = 0;
        while (count > 0) {
            var random = Math.floor(Math.random() * count);
            var obj = transformList[random];
            transformList.splice(random, 1);
            new Game.Timer("skill", i*f, 1, delegate(this, SpecialAnimationPlayer.onLateActiveChildren, obj));
            count--;
            i++;
        }
    }
}
export default SpecialAnimationPlayer;

File: publish/project/Assets/Third/ScriptCommon/Anim/SpecialAnimationPlayer.cs
using UnityEngine;
using System.Collections.Generic;

public class SpecialAnimationPlayer : MonoBehaviour
{
    public string functionName = null;
    public string[] funcParams = null;
#if PLAY_CS
    private System.Action lateCall = null;
    // Use this for initialization
    void Awake()
    {
        this.Invoke(functionName,0);
    }
    void ActiveChildren()
    {
        var count = transform.childCount;
        foreach (Transform t in transform)
        {
            t.gameObject.SetActive(false);
        }
        if (count == 0)
        {
            return;
        }
        this.Invoke("LateCall", 0);
        int index = 0;
        lateCall = () =>
        {
            transform.GetChild(index).gameObject.SetActive(true);
            index++;
            if (index < count)
            {
                this.Invoke("LateCall", GetFloat(0));
            }
            else
            {
                lateCall = null;
            }
        };
    }
    void RandomActiveChildren()
    {
        List<Transform> transformList = new List<Transform>();
        foreach (Transform t in transform)
        {
            t.gameObject.SetActive(false);
            transformList.Add(t);
        }
        if (transformList.Count == 0)
        {
            return;
        }
        this.Invoke("LateCall", 0);
        lateCall = () =>
        {
            var random = Random.Range(0, transformList.Count);

            transformList[random].gameObject.SetActive(true);
            transformList.RemoveAt(random);
            if (transformList.Count > 0)
            {
                this.Invoke("LateCall", GetFloat(0));
            }
            else
            {
                lateCall = null;
            }
        };
    }



    void LateCall()
    {
        if(lateCall!=null)
            lateCall();
    }
#endif


    public double GetDouble(int index)
    {
        if (funcParams == null || funcParams.Length <= index)
        {
            return 0;
        }
        double value = 0;
        double.TryParse(funcParams[index], out value);
        return value;
    }
    public float GetFloat(int index)
    {
        if (funcParams == null || funcParams.Length <= index)
        {
            return 0;
        }
        float value = 0;
        float.TryParse(funcParams[index], out value);
        return value;
    }

    public string GetString(int index)
    {
        if (funcParams == null || funcParams.Length <= index)
        {
            return "";
        }
        return funcParams[index];
    }
}